Refactor KubeVirtCSIDriver creation

alifelan opened the pull request:
The logic and dependencies on the creation of the Controller and the Node services differ. Splitting them will make them more manageable.

While trying to keep the functionality, there are a few key differences:
1. A nodeName is required to run the Node service
2. If both the Controller and Node service are being run from the cmd binary, it will fallback to the identity controller in the tenant clientset. As far as I understand, running both services from the same binary (as in, within the same pod) is not a valid scenario, with the only instance i found being in the sanity test.

I'll follow with some testing.

**What this PR does / why we need it**:
Refactors KubeVirtCSIDriver creation and KubeVirtCSI binary entry point. Separates Controller and Node service creation to facilitate updates.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This started off from introducing informers and a fix to RWO volumes having two different volume attachments at the same time. There will be follow ups to cover those linked to this PR. While working on that, I noticed that a refactor may be useful.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
4. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alifelan committed:
Refactor KubeVirtCSIDriver creation

The logic and dependencies on the creation of the Controller and the
Node services differ. Splitting them will make them more manageable.

Change-Id: I4cc977a8cacbc05473a7e8ef00b6f5be0e391992
Signed-off-by: Alí Felán <alifelan@google.com>

alifelan committed:
Create config struct and migrate to using flagsets

This introduces a config struct to hold parameters passed to the
kubevirt-csi-driver command, enabling passing those parameters to other
functions without relying on global variables. Along this, we also
migrate to using flagsets, separating our parameters from a global
definition, which may only be parsed once.

Change-Id: Ic54058f76ee053e2a83bed3fae0400c91f5d200b
Signed-off-by: Alí Felán <alifelan@google.com>

alifelan committed:
Move all exits to the main function

We now return errors where appropriate instead of calling an exit from
the program. All exists are handled in main, allowing us to run unit
tests for all utility functions.

Change-Id: Icaa4ac216811e03ef05fc6374cee70d4f2ef1c10
Signed-off-by: Alí Felán <alifelan@google.com>

alifelan committed:
Split prechecks for services

Instead of checking for prechecks initially, we are now checking the
values when needed. This allows each service to handle only the checks
that it needs.

Change-Id: Ie6533920bcf6330c864750aa568d7593133c8ee4
Signed-off-by: Alí Felán <alifelan@google.com>

alifelan committed:
Move rest config and clienset management to config struct

Allows us to pass clients in unit tests.

Change-Id: I7d52ab7df09e279d605bb5032445bbd48bc26176
Signed-off-by: Alí Felán <alifelan@google.com>

akalenyu commented:
/test pull-csi-driver-e2e-k8s

akalenyu commented:
```
Waiting for deployment "kubevirt-csi-controller" rollout to finish: 0 of 1 updated replicas are available...
error: timed out waiting for the condition
```

probably unrelated

alifelan committed:
Add error logging to driver rollouts

This now logs details on the failed resource and its associated pods.

Change-Id: I3cccf43de6293120a5f17647f7655cc10b249644
Signed-off-by: Alí Felán <alifelan@google.com>

alifelan commented:
hey @akalenyu! I tried recreating locally, but it passed. I'm not sure whats happening here, but I've added some error handling to the rollout status command so we get some details when needed. What do you think of that? Can we retry the tests with this?

alifelan commented:
While testing locally i ran into an error in `make cluster-sync`, which is the same as the deployment not being healthy. Im adding how the output of the failure looks: [make-cluster-sync.log](https://github.com/user-attachments/files/25051689/make-cluster-sync.log)

The tldr is that its trying to start the node service in the controller. From what i see, the flag has always been defaulted to true (before and after this refactor), so theres some other logic i modified to cause this
```
alifelan_google_com@alifelan-kvcsi-test:~/kubevirt.io/csi-driver$ ./kubevirtci kubectl-tenant get pod -n kubevirt-csi-driver kubevirt-csi-controller-56ffcf779f-m9kwj
selecting docker as container runtime
Found control plane VM: kvcluster-control-plane-9bwfs in namespace kvcluster
NAME                                       READY   STATUS             RESTARTS        AGE
kubevirt-csi-controller-56ffcf779f-m9kwj   2/6     CrashLoopBackOff   21 (107s ago)   9m48s
alifelan_google_com@alifelan-kvcsi-test:~/kubevirt.io/csi-driver$ ./kubevirtci kubectl-tenant logs -n kubevirt-csi-driver kubevirt-csi-controller-56ffcf779f-m9kwj -c csi-driver
selecting docker as container runtime
Found control plane VM: kvcluster-control-plane-9bwfs in namespace kvcluster
I0203 16:24:01.276429       1 kubevirt-csi-driver.go:78] Driver vendor csi.kubevirt.io 0.2.0
I0203 16:24:01.276820       1 kubevirt-csi-driver.go:108] Storage class enforcement string:
F0203 16:24:01.280368       1 kubevirt-csi-driver.go:32] Failed to initialize driver: failed to configure node service: cannot start NodeService without a node name.
```

Ill look into it. What i find interesting is that the [latest run](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_csi-driver/159/pull-csi-driver-e2e-k8s/2018614124623171584) had the controller pass the rollout status. Ill leave a rollout status running on the background on my side while i check what happened with the node service

alifelan committed:
Explicitly set services on controller-tenant

We were relying on default values, which assume true for either of the
services. With the new refactor, we are now more strict on the
deployment. Now, we disable the node service in the controller

Change-Id: Ic82fda0d4a548df74d1a5b76e1529ac44ef6027a
Signed-off-by: Alí Felán <alifelan@google.com>

awels commented:
/test all

awels reviewed (approved):
/approve
